logger.js

```diff
@@ -0,0 +1,116 @@
+function log(message) {
+    console.log(message);
+}
+
+function log2(...msgArgs) {
+    let message = msgArgs.join(" | ");
+    log(message);
+}
+
+function log3(...msgArgs) {
+    let date = new Date();
+    let dateStr = `${date.toLocaleDateString()}, ${date.toLocaleTimeString()}`;
+    log2(dateStr, ...msgArgs);
+}
+
+log("Hello World !");
+log2("Hello", "World", "!");
+log3("Hello", "World", "!");
+
+
+//Functional pattern
+
+let baseLogger = {};
+(function (context) {
+    context.log = function (message) {
+        console.log(message);
+    };
+})(baseLogger);
+
+let tableLogger = {};
+tableLogger.log = function (...msgArgs) {
+    let message = msgArgs.join(" | ");
+    baseLogger.log(message);
+};
+
+let tableTimeLogger = {};
+tableTimeLogger.log = function (...msgArgs) {
+    let date = new Date();
+    let dateStr = `${date.toLocaleDateString()}, ${date.toLocaleTimeString()}`;
+    tableLogger.log(dateStr, ...msgArgs);
+};
+
+baseLogger.log("Hello World !");
+tableLogger.log("Hello", "World", "!");
+tableTimeLogger.log("Hello", "World", "!");
+
+
+//Class inheritance pattern
+
+class BaseLogger {
+    log(message) {
+        console.log(message);
+    };
+}
+
+class TableLogger extends BaseLogger {
+    log(...msgArgs) {
+        let message = msgArgs.join(" | ");
+        super.log(message);
+    };
+}
+
+class TableTimeLogger extends TableLogger {
+    log(...msgArgs) {
+        let date = new Date();
+        let dateStr = `${date.toLocaleDateString()}, ${date.toLocaleTimeString()}`;
+        super.log(dateStr, ...msgArgs);
+    }
+}
+
+baseLogger = new BaseLogger();
+baseLogger.log("Hello World !");
+
+tableLogger = new TableLogger();
+tableLogger.log("Hello", "World", "!");
+
+tableTimeLogger = new TableTimeLogger();
+tableTimeLogger.log("Hello", "World", "!");
+
+
+//Prototypical pattern
+
+function BaseLogger2() {
```

Comment:
If you split your code into a separate files I don't have to name your classes with counters